docker-images/kafka/scripts/kafka_connect_run.sh

```diff
@@ -54,4 +54,4 @@ fi
 . ./set_kafka_gc_options.sh
 
 # starting Kafka server with final configuration
-exec $KAFKA_HOME/bin/connect-distributed.sh /tmp/strimzi-connect.properties
+/usr/bin/tini -w -e 143 -- sh -c "${KAFKA_HOME}/bin/connect-distributed.sh /tmp/strimzi-connect.properties"
```

Comment:
So we're not actually running tini as PID 1 then?

Reply:
That's a very good point. 

When we've done this before, we've made the tini command the entrypoint for the Docker image (e.g. see https://hub.docker.com/layers/ibmcom/eventstreams-kafka-ce-icp-linux-amd64/2019.4.1-095ae1e/images/sha256-29e3b7dc465ffdbe21bf76d3a33397f05a709eca376a7cd90326b9905b2c85d2 ). Because there was no equivalent place to do it in the Dockerfile here, we defaulted to this as the quickest/easiest place to insert tini. 

But I think you're right, that this wasn't valid. 

We could try running tini in a mode that doesn't depend on being PID 1 (adding the `-s` option) but I'm not as fond of that. 

What do you think about inserting it in the Java container definitions instead? (e.g. https://github.com/strimzi/strimzi-kafka-operator/blob/eb85fafffc72aeefa4a25ea3b30425a8bc718f60/cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaCluster.java#L1399 )

Reply:
Having a start up script (i.e. `bash`) call `tini` to run another startup script (as you have above) means that bash is (and remains, because you haven't `exec tini`) PID 1 and will e.g. do the wrong thing with signals.

I guess you could change what you have above to `exec tini ...`, in which case bash would initially be PID 1, but be replaced by tini running initially another bash and later the jvm (when the start up script calls `exec $JAVA`). 

If we changed `KafkaCluster` to use command `tini /opt/kafka/kafka_run.sh` then tini would run as pid 1 and execute bash which would interpret the startup scripts eventually reaching `exec $JAVA` in `kafka-run-class.sh` which would replace the shell with the jvm (without starting a new process). I think that's what you want.

But the only way to test it will be to `kubectl exec` into the pod and look at the process tree for real.

Reply:
Having read https://github.com/krallin/tini/issues/8 I don't really understand what tini is doing for us. I can see it would be useful _if_ a component was starting subprocesses from the JVM, but generally components don't do that. And I can see it would be useful if bash was intercepting signals, but we generally `exec java`, so bash is removed from the picture and signals get sent directly to the JVM, which (AFAIK) has its signal handlers set up in a more or less sane way.

Can you elaborate on what problems tini would solve?

Reply:
Hi @tombentley we have addressed the issue of tini not running as pid 1 by using exec as previously mentioned, and added it to `launch_java` to ensure the operator also runs with tini (also as pid 1).

With regards to the benefit of adding tini, we believe it will solve these problems:
- Although Java does indeed have signal handling, it does not handle the cleanup of zombie processes which tini does do
- Java's signal handling will result in a 143 exit code upon receiving a `docker stop`, kubernetes interprets this as an "error" state. Tini allows us to map this code to a 0 (via `-e 143`), which kubernetes interprets as being a graceful termination of a pod and displays no such "error" state
- We noticed the OpenSslCertManager in the operator does appear to be running process execs which would be helped by tini's cleanup ability
- Currently, we don't believe that Java is running any process execs, however, should it in the future we would be protected from zombie processes thanks to tini.
- Any exec probes (kubernetes liveness/readiness) that are run in a container are parented to whatever is pid 1 so having tini in-place as pid 1 will correctly deal with the lifecycle of those (signal handling / zombies etc.) whereas if java is left as pid 1 it wouldn't do so

Since tini benefits the operator, it makes sense for it to be applied consistently to other Java components.

Reply:
@tombentley Does this sound reasonable?

Reply:
Yes, I think it does sound reasonable (I'd forgotten about `OpenSslCertManager` for one thing, and I didn't know about the 143 status code).